### DISCUSSION TITLE
Defining a @Query with a variable

### DISCUSSION THREAD

**Discussion by foodstan** [groue/GRDBQuery#29] [Q&A] [answered]
Hi,

I have a `NavigationView` listing groups, then when tapping on one it switches to a list of messages for that group, and then tapping on a message will load a view of that specific message. 

The main `@Query` finding all groups works fine in `MainView`, but how do I make a message `@Query` in the `MessageList` that only includes messages associated with the `group`?  It won't allow me to pass in the local `group` variable because it has to be done at init time.

The views:

```
struct MainView: View {
    @Environment(\.appDatabase) private var appDatabase
    @Query(GroupRequest(ordering: .byName)) private var groups: [Group]

    var body: some View {
      NavigationView {
        List {
          ForEach(groups) { group in
            NavigationLink(destination: MessageList(group: group)) {
              GroupRow(group: group)
            }
          }
        }
      }
   }
}

struct MessageList: View {
    @Environment(\.appDatabase) private var appDatabase
    var group: Group?

    // @Query(MessageRequest(group: group)) private var messages: [Message]
    // error: Cannot use instance member 'group' within property initializer; property initializers run before 'self' is available

    var messages: [Message] {
        do {
            return try appDatabase.databaseReader.read { db in
                return try group!.messages.fetchAll(db)
            }
        } catch {
        }
	return []
    }

    var body: some View {
      List {
        ForEach(messages) { message in
          NavigationLink(destination: MessageView(message: message)) {
            MessageRow(message: message)
          }
        }
      }
    }
}
```

This works, but it's not a `@Query` in `MessageList` so when new messages are inserted, they don't automatically appear in the list like they do from GroupView.

How can I construct a `@Query` in `MessageList` that uses the `group` passed in from the `NavigationLink`?

**Comment by groue** [comment, marked as answer]
Hello @foodstan,

Check the "Configuring the Initial Request" section of [Adding Parameters to Queryable Types](https://groue.github.io/GRDBQuery/0.5/documentation/grdbquery/queryableparameters), and the following sections.

In your case, since `MainView` just wants to initialize MessageList with an "initial request", you'll probably end up with:

```swift
struct MessageList: View {
    @Query<MessageRequest> private var messages: [Message]

    init(group: Group) {
        let request = MessageRequest(group: group)
        _messages = Query(request)
    }
}
```

This will allow `MessageList` to further control the request if it needs (filter messages, etc), through `$messages`.

**Reply by foodstan** [reply]
Thanks!
